l3kernel/l3file.dtx

```diff
@@ -2849,9 +2849,40 @@
 \cs_new:Npn \@@_full_name:n #1
   {
     \tl_if_blank:nF {#1}
-      { \exp_args:Nne \@@_full_name_aux:nn {#1} { \@@_size:n {#1} } }
+      { \exp_args:Nne \@@_full_name_auxii:nn {#1} { \@@_full_name_aux:n {#1} } }
   }
-\cs_new:Npn \@@_full_name_aux:nn #1 #2
+%    \end{macrocode}
+%   To avoid repeated reading of files we need to cache the loading:
+%   this is important as the code here is used by \emph{all} file checks.
+%   The same marker is used in the \LaTeXe{} kernel, meaning that we get a
+%   double-saving with for example \cs{IfFileExists}. As this is all about
+%   performance, we use the low-level approach for the conditionals. For
+%   a file already seen, the size is reported as $-1$ so it's distinct from
+%   any non-cached ones.
+%    \begin{macrocode}
+\cs_new:Npn \@@_full_name_aux:n #1
+  {
+    \if_cs_exist:w @@_seen_ \tl_to_str:n {#1} : \cs_end:
+      -1
+    \else:
+      \exp_args:Ne \@@_full_name_auxi:nn { \@@_size:n {#1} } {#1}
+    \fi:
+  }
+%    \end{macrocode}
+%   We will need the size of files later, and we have to avoid the
+%   \cs{scan_stop:} causing issues if we are raising the flag. Thus there is
+%   a slightly odd gobble here.
+%    \begin{macrocode}
+\cs_new:Npn \@@_full_name_auxi:nn #1#2
+  {
+    \if:w \scan_stop: #1 \scan_stop:
+    \else:
+      \exp_after:wN \use_none:n
+        \cs:w @@_seen_ \tl_to_str:n {#2} : \cs_end:
+    \fi:
+    #1
```

Comment:
not sure I understand that one. If `#1` aka `\@@_size:n{...}` returns empty then why execute it a second time? I.e., shouldn't the `#1` move into the `\else:`?

Reply:
I've already `e`-type expanded `#1`, so it is the result of `\pdffilesize`; thus it's either the value we need to return *or* it's entirely empty. Either way, it doesn't need to be inside the conditional.

Reply:
ah yes, it is already gone, so no further file access, but still a tiny bit slower, because TeX has to look up what is at `#1` (nothing) instead of just bypassing the else clause at high speed --- most likely never measurable though